### PR TITLE
python3Packages.nvidia-ml-py: init 11.515.48

### PR DESCRIPTION
**Pull request by GaetanLepage**
###### Description of changes

Adding the `nvidia-ml-py` bindings for the NVIDIA Management Library.
This is a requirement for [nvitop](https://github.com/XuehaiPan/nvitop#installation) which I would like to package next.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Hello @FRidh and @jonringer  !

I think this PR should be ready.
Please, whenever you have the time, could you have a look at it ?
I am of course willing to address any necessary modification.

Thank you very much in advance :)

**Comment by FRidh** [comment]
Note quite.
```
nvmlLib = CDLL("libnvidia-ml.so.1")
```
needs to hardcode the absolute path to the library.

**Comment by SomeoneSerge** [comment]
I guess we can preserve `LD_LIBRARY_PATH` behaviour if we try paths in this order:

```nix
''
try:
    nvmlLib = CDLL("libnvidia-ml.so.1")
except OSError:
    nvmlLib = CDLL("${addOpenGLRunpath.driverLink}")
''
```

Idk if it affects nixGL in any way

**Comment by GaetanLepage** [comment]
@SomeoneSerge is your suggestion supposed to be a patch ?
I don't really see what I should change...
Thank you anyway for your help !

**Comment by SomeoneSerge** [comment]
@GaetanLepage 

There's the line `1679` in `pynvml.py` that @ FRidh mentioned:
```python
                         nvmlLib = CDLL("libnvidia-ml.so.1")
```

If you run this on NixOS it's going to fail with `OSError`, because the directory with `libnvidia-ml.so.1` won't be in the search path list. I suggest we catch this error and re-try with this library's absolute path. This library seems to be driver-specific (like `libcuda.so`), so instead of a path in `/nix/store` we probably should use a path in `/run/opengl-driver/lib`, deployed by NixOS at runtime

File: pkgs/development/python-modules/nvidia-ml-py/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+, fetchPypi
+, buildPythonPackage
+}:
+
+buildPythonPackage rec {
+  pname = "nvidia-ml-py";
+
+  version = "11.515.48";
+
+  format = "setuptools";
```
Comment (jonringer) [inline pkgs/development/python-modules/nvidia-ml-py/default.nix:11]:
```suggestion
  pname = "nvidia-ml-py";
  version = "11.515.48";
  format = "setuptools";
```

File: pkgs/development/python-modules/nvidia-ml-py/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    extension = "tar.gz";
+    hash = "sha256-iNLQu9c8Q3B+FXMObRTtxqE3B/siJIlIlCH6T0rX+sY=";
+  };
+
+  nativeBuildInputs = [ ];
+  buildInputs = [ ];
+
```
Comment (jonringer) [inline pkgs/development/python-modules/nvidia-ml-py/default.nix:21]: if they're not being used, just remove
```suggestion
```

**Review by jonringer — COMMENTED** [review]
as others have mentioned, if this is exposing the functionality of nvidia drivers, you need to have the `CDLL` call resolve to nixos driver path `/run/opengl-driver/lib/<lib>`

**Comment by SuperSandro2000** [comment]
Missing tests or doCheck = false and pythonImportsCheck.

**Comment by FRidh** [comment]
Could we not fix this in the stub and have all packages patch the path to the stub instead?

**Comment by GaetanLepage** [comment]
Am I then supposed to use `substituteInPlace` to replace `the string `libnvidia-ml.so.1` by `/run/opengl-driver/lib/libnvidia-ml.so.1` in `pynvml.py` or is there something smarter to do ?

**Comment by SomeoneSerge** [comment]
> Could we not fix this in the stub and have all packages patch the path to the stub instead?

I'm not sure I follow, what exactly are you suggesting?
Trying to use e.g. `nvidia-smi` linked against the stub won't work

> substituteInPlace to replace the string libnvidia-ml.so.1 by /run/opengl-driver/lib/libnvidia-ml.so.1

It's crude, but that's what i was thinking of

File: pkgs/development/python-modules/nvidia-ml-py/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  version = "11.515.48";
+  format = "setuptools";
+
+  src = fetchPypi {
+    inherit pname version;
+    extension = "tar.gz";
+    hash = "sha256-iNLQu9c8Q3B+FXMObRTtxqE3B/siJIlIlCH6T0rX+sY=";
+  };
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nvidia-ml-py/default.nix:17]:
```suggestion
  # no tests
  doCheck = false;
```

**Comment by GaetanLepage** [comment]
> > Could we not fix this in the stub and have all packages patch the path to the stub instead?
> 
> I'm not sure I follow, what exactly are you suggesting? Trying to use e.g. `nvidia-smi` linked against the stub won't work
> 
> > substituteInPlace to replace the string libnvidia-ml.so.1 by /run/opengl-driver/lib/libnvidia-ml.so.1
> 
> It's crude, but that's what i was thinking of

done

**Comment by FRidh** [comment]
> I'm not sure I follow, what exactly are you suggesting?
> Trying to use e.g. `nvidia-smi` linked against the stub won't work

Ah okay, I thought that would be possible. Refer from here to the stub, and then in the stub refer to the driver library.

**Comment by SomeoneSerge** [comment]
✅  Tested the runtime with

```nix
# shell.nix
let
  nixpkgs = builtins.getFlake github:GaetanLepage/nixpkgs/nvidia-ml-py;
  pkgs = import nixpkgs { config.cudaSupport = true; config.allowUnfree = true; };
  py = pkgs.python3.withPackages (ps: with ps; [
    nvidia-ml-py
  ]);
in
with pkgs;
mkShell {
  packages = [ py ];
}
```

and 

```python
# test.py
import pynvml

if __name__ == "__main__":
    pynvml.nvmlInit()

    print(pynvml.nvmlSystemGetDriverVersion())

    n_devices = pynvml.nvmlDeviceGetCount()
    devices = [
        pynvml.nvmlDeviceGetName(pynvml.nvmlDeviceGetHandleByIndex(i))
        for i in range(n_devices)
    ]
    print(f"{n_devices} devices:")
    print("\n".join(devices))
```

**Comment by GaetanLepage** [comment]
Can the `libnvidia-ml.so.1` hard linking as `/run/opengl-driver/lib/libnvidia-ml.so.1` pose some problems on non-NixOs systems ?

**Comment by FRidh** [comment]
> Can the `libnvidia-ml.so.1` hard linking as `/run/opengl-driver/lib/libnvidia-ml.so.1` pose some problems on non-NixOs systems ?

Yes, this Nix package with that patch will only be usable on NixOS.

**Comment by SomeoneSerge** [comment]
> Can the libnvidia-ml.so.1 hard linking as /run/opengl-driver/lib/libnvidia-ml.so.1 pose some problems on non-NixOs systems ?

This why I suggested a try-except patch instead: with it non-nixos users can set `LD_LIBRARY_PATH=...` as a workaround

**Comment by SomeoneSerge** [comment]
I don't find any of this mandatory, but if it makes sense for you, I meant something like:

```diff
diff --git a/pynvml.py b/pynvml.py
index 9e094c6..727716b 100644
--- a/pynvml.py
+++ b/pynvml.py
@@ -1676,7 +1676,11 @@ def _LoadNvmlLibrary():
                             nvmlLib = CDLL(os.path.join(os.getenv("ProgramFiles", "C:/Program Files"), "NVIDIA Corporation/NVSMI/nvml.dll"))
                     else:
                         # assume linux
-                        nvmlLib = CDLL("libnvidia-ml.so.1")
+                        try:
+                            nvmlLib = CDLL("libnvidia-ml.so.1")
+                        except OSError:
+                            # Assume NixOS
+                            nvmlLib = CDLL("/run/opengl-driver/lib/libnvidia-ml.so.1")
                 except OSError as ose:
                     _nvmlCheckReturn(NVML_ERROR_LIBRARY_NOT_FOUND)
                 if (nvmlLib == None):
```


I would, maybe, define this patch in Nix (via `writeText`?), and use the `addOpenGLRunpath.driverLink` value instead of `/run/opengl-driver/...` because there's an intent to rename the "opengl-driver" eventually

**Comment by GaetanLepage** [comment]
> I would, maybe, define this patch in Nix (via `writeText`?), and use the `addOpenGLRunpath.driverLink` value instead of `/run/opengl-driver/...` because there's an intent to rename the "opengl-driver" eventually.
> 
> Maybe it would be even simpler to substitute `/run/opengl-driver/lib` with `driverLink` after the patch phase (the substitute would require a comment explaining that it's there in case of a rename, and the CDLL line in the diff would require a comment that this gets substituted later)

I am not sure how to properly do this...
Do I need to create a separate `libnvidia-path.patch` file and reference it in the derivation ?

**Comment by SomeoneSerge** [comment]
```nix
buildPythonPackage {
    # ...
    patches = [
        ./0001-locate-libnvidia-ml.so.1-on-NixOS.patch
    ];
    # ...
}
```
or
```nix
buildPythonPackage {
    # ...
    patches = [
        (writeText "0001-locate-libnvidia-ml.so.1-on-NixOS.patch" ''
            ...
        '')
    ];
    # ...
}
```

**Comment by GaetanLepage** [comment]
Ok, I have adopted the patch strategy.
It should be ok now.

@SomeoneSerge could you please test on your system please ?

**Comment by SomeoneSerge** [comment]
- [x] Works on NixOS
- [ ] Haven't tested `LD_LIBRARY_PATH=...` but I expect it work
- [ ] Last thing to choose is whether we account for the value of `addOpenGLRunpath.driverLink` right now, or later in https://github.com/NixOS/nixpkgs/pull/158079

**Comment by SuperSandro2000** [comment]
> or
> 
> ```nix
> buildPythonPackage {
>     # ...
>     patches = [
>         (writeText "0001-locate-libnvidia-ml.so.1-on-NixOS.patch" ''
>             ...
>         '')
>     ];
>     # ...
> }
> ```

Please don't do that. It needlessly creates a derivation.

File: pkgs/development/python-modules/nvidia-ml-py/default.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  };
+
+  # no tests
+  doCheck = false;
+
+  pythonImportsCheck = [ "pynvml" ];
+
+  patches = [
+    ./0001-locate-libnvidia-ml.so.1-on-NixOS.patch
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nvidia-ml-py/default.nix:24]:
```suggestion
```

File: pkgs/development/python-modules/nvidia-ml-py/default.nix [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  pname = "nvidia-ml-py";
+  version = "11.515.48";
+  format = "setuptools";
+
+  src = fetchPypi {
+    inherit pname version;
+    extension = "tar.gz";
+    hash = "sha256-iNLQu9c8Q3B+FXMObRTtxqE3B/siJIlIlCH6T0rX+sY=";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/nvidia-ml-py/default.nix:16]:
```suggestion

  patches = [
    ./0001-locate-libnvidia-ml.so.1-on-NixOS.patch
  ];

```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/python-modules/nvidia-ml-py/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  patches = [
+    ./0001-locate-libnvidia-ml.so.1-on-NixOS.patch
+  ];
+
+  meta = {
+    description = "Python Bindings for the NVIDIA Management Library";
+    homepage = "https://pypi.org/project/nvidia-ml-py";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ GaetanLepage ];
```
Comment (FRidh) [inline pkgs/development/python-modules/nvidia-ml-py/default.nix:30]: might want to set platforms as this package cannot work on darwin. Should probably be x86_64-linux only.

Comment (GaetanLepage) [reply pkgs/development/python-modules/nvidia-ml-py/default.nix:31]: Done :)

**Review by FRidh — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.nvidia-ml-py

**Pull request merged**